### DISCUSSION TITLE
Software trigger mode in python environment

### DISCUSSION THREAD

**Discussion by Astronomia-maker** [ifm/ifm3d#386] [Q&A]
Hi, I am using two O3R camera with VPU. Both camera runs parallel for grabbing an image and 3D data. So i want to configure both camera on software trigger mode. But i am getting always stuck at frame= await fg.wait_for_frame().

I have installed ifm3dpy==1.1.1 version. Here is my whole code with configuration. Can you help me to set both camera for software trigger mode.
from ifm3dpy import O3R, FrameGrabber, buffer_id
import asyncio
o3r = O3R('10.100.100.31')

o3r.set(
    {
    "applications": {
        "classes": {
            "ods": {
                "instanceSchema": {},
                "license": "UNLICENSED",
                "minimumFirmwareVersion": {
                    "major": 0,
                    "minor": 16,
                    "patch": 3
                },
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            }
        }
    },
    "device": {
        "clock": {
            "currentTime": 1581090712488271648,
            "sntp": {
                "active": True,
                "activeServerAddress": "",
                "activeServerName": "time1.google.com",
                "availableServers": [],
                "systemClockSynchronized": False
            }
        },
        "diagnostic": {
            "confInitStages": [
                "device",
                "ports",
                "applications"
            ],
            "temperatures": [
                {
                    "entity": "PLL-therm",
                    "value": 31.0
                },
                {
                    "entity": "MCPU-therm",
                    "value": 31.0
                },
                {
                    "entity": "PMIC-Die",
                    "value": 100.0
                },
                {
                    "entity": "Tboard_tegra",
                    "value": 26.0
                },
                {
                    "entity": "GPU-therm",
                    "value": 28.0
                },
                {
                    "entity": "BCPU-therm",
                    "value": 31.0
                },
                {
                    "entity": "thermal-fan-est",
                    "value": 29.5
                },
                {
                    "entity": "Tdiode_tegra",
                    "value": 26.75
                }
            ],
            "upTime": 68000000000
        },
        "info": {
            "description": "",
            "device": "0302",
            "deviceTreeBinaryBlob": "tegra186-quill-p3310-1000-c03-00-base.dtb",
            "features": {},
            "guiSettings": "",
            "name": "",
            "partNumber": "OVP800",
            "productionState": "AA",
            "serialNumber": "000020228630",
            "vendor": "0001"
        },
        "log": {
            "components": {},
            "level": "warn",
            "storage": "volatile"
        },
        "network": {
            "authorized_keys": "",
            "interfaces": {
                "eth0": {
                    "ipAddressConfig": 0,
                    "ipv4": {
                        "address": "10.100.100.31",
                        "dns": "255.255.255.0",
                        "gateway": "10.100.100.1",
                        "mask": 24
                    },
                    "mac": "48:B0:2D:55:03:AC",
                    "networkSpeed": 1000,
                    "useDHCP": False
                },
                "eth1": {
                    "ipAddressConfig": 0,
                    "mac": "00:02:01:43:70:24",
                    "networkSpeed": 0,
                    "useDHCP": True
                }
            }
        },
        "status": "OPERATE",
        "swVersion": {
            "euphrates": "1.18.3",
            "firmware": "0.16.23-823",
            "kernel": "4.9.140-l4t-r32.4+g8c7b68130d9a",
            "l4t": "r32.4.3",
            "schema": "v1.4.1",
            "tcu": "1.0.4"
        }
    },
    "ports": {
        "port0": {
            "acquisition": {
                "delay": 0,
                "framerate": 10.0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "RGB_INFO"
                ],
                "pcicTCPPort": 50010
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 1,
                        "patch": 0
                    }
                },
                "device": "2301",
                "deviceTreeBinaryBlobOverlay": "001-ov9782.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 127,
                        "vertical": 80
                    },
                    "resolution": {
                        "height": 800,
                        "width": 1280
                    },
                    "type": "2D"
                },
                "icc": {
                    "version": {
                        "major": 0,
                        "minor": 0,
                        "patch": 0
                    }
                },
                "name": "",
                "partNumber": "M03933",
                "productionState": "AA",
                "sensor": "OV9782",
                "sensorID": "OV9782_127x80_noIllu_Csample",
                "serialNumber": "000000000812",
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_autoexposure2D",
            "processing": {
                "extrinsicHeadToUser": {
                    "rotX": 0.45,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": -0.15,
                    "transY": 1.74,
                    "transZ": -0.07
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        },
        "port1": {
            "acquisition": {
                "delay": 0,
                "framerate": 10.0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "RGB_INFO"
                ],
                "pcicTCPPort": 50011
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 1,
                        "patch": 0
                    }
                },
                "device": "2301",
                "deviceTreeBinaryBlobOverlay": "001-ov9782.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 127,
                        "vertical": 80
                    },
                    "resolution": {
                        "height": 800,
                        "width": 1280
                    },
                    "type": "2D"
                },
                "icc": {
                    "version": {
                        "major": 0,
                        "minor": 0,
                        "patch": 0
                    }
                },
                "name": "",
                "partNumber": "M03933",
                "productionState": "AA",
                "sensor": "OV9782",
                "sensorID": "OV9782_127x80_noIllu_Csample",
                "serialNumber": "000000000809",
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_autoexposure2D",
            "processing": {
                "extrinsicHeadToUser": {
                    "rotX": 0.0,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": 0.0,
                    "transY": 0.635,
                    "transZ": 0.0
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        },
        "port2": {
            "acquisition": {
                "channelSelection": "manual",
                "channelValue": 2,
                "delay": 0,
                "exposureLong": 1200,
                "exposureShort": 1000,
                "framerate": 15.0,
                "offset": 0.0,
                "swTriggerGroup": 2,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "TOF_INFO",
                    "RADIAL_DISTANCE_NOISE",
                    "RADIAL_DISTANCE_COMPRESSED",
                    "REFLECTIVITY",
                    "AMPLITUDE_COMPRESSED",
                    "CONFIDENCE"
                ],
                "pcicTCPPort": 50012
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 4,
                        "patch": 2
                    }
                },
                "device": "3101",
                "deviceTreeBinaryBlobOverlay": "001-irs2381c.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 60,
                        "vertical": 45
                    },
                    "resolution": {
                        "height": 172,
                        "width": 224
                    },
                    "type": "3D"
                },
                "icc": {
                    "version": {
                        "major": 1,
                        "minor": 0,
                        "patch": 6
                    }
                },
                "name": "",
                "partNumber": "M03933",
                "productionState": "AA",
                "sensor": "IRS2381C",
                "sensorID": "IRS2381C_60x45_4x2W_60x45_C2",
                "serialNumber": "000000000812",
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_range4m",
            "processing": {
                "diParam": {
                    "anfFilterSizeDiv2": 2,
                    "dynamicSymmetryThreshold": 20.0,
                    "enableTemporalFilter": True,
                    "maxDistNoise": 0.05,
                    "medianSizeDiv2": 0,
                    "minAmplitude": 30.0,
                    "minReflectivity": 3.0,
                    "mixedPixelFilterMode": 1,
                    "mixedPixelThresholdRad": 0.05
                },
                "extrinsicHeadToUser": {
                    "rotX": 0.45,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": -0.15,
                    "transY": 1.7,
                    "transZ": -0.15
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "IDLE"
        },
        "port3": {
            "acquisition": {
                "channelSelection": "manual",
                "channelValue": 4,
                "delay": 0,
                "exposureLong": 3000,
                "exposureShort": 1000,
                "framerate": 20.0,
                "offset": 0.0,
                "swTriggerGroup": 3,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "TOF_INFO",
                    "RADIAL_DISTANCE_NOISE",
                    "RADIAL_DISTANCE_COMPRESSED",
                    "REFLECTIVITY",
                    "AMPLITUDE_COMPRESSED",
                    "CONFIDENCE"
                ],
                "pcicTCPPort": 50013
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 4,
                        "patch": 2
                    }
                },
                "device": "3101",
                "deviceTreeBinaryBlobOverlay": "001-irs2381c.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 60,
                        "vertical": 45
                    },
                    "resolution": {
                        "height": 172,
                        "width": 224
                    },
                    "type": "3D"
                },
                "icc": {
                    "version": {
                        "major": 1,
                        "minor": 0,
                        "patch": 6
                    }
                },
                "name": "",
                "partNumber": "M03933",
                "productionState": "AA",
                "sensor": "IRS2381C",
                "sensorID": "IRS2381C_60x45_4x2W_60x45_C2",
                "serialNumber": "000000000809",
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_range4m",
            "processing": {
                "diParam": {
                    "anfFilterSizeDiv2": 2,
                    "dynamicSymmetryThreshold": 20.0,
                    "enableTemporalFilter": True,
                    "maxDistNoise": 0.05,
                    "medianSizeDiv2": 0,
                    "minAmplitude": 30.0,
                    "minReflectivity": 3.0,
                    "mixedPixelFilterMode": 1,
                    "mixedPixelThresholdRad": 0.5
                },
                "extrinsicHeadToUser": {
                    "rotX": 0.0,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": 0.0,
                    "transY": 0.638,
                    "transZ": 0.0
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "IDLE"
        }
    }
}
)
o3r.save_init()

async def bottom():
  fg = FrameGrabber(o3r, pcic_port=50012)  
  fg.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
  fg.sw_trigger()
  frame = await fg.wait_for_frame()
  cartesian_3d = frame.get_buffer(buffer_id.XYZ)
  fg.stop()
  
async def upper():

  fg = FrameGrabber(o3r, pcic_port=50013)  
  fg.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
  fg.sw_trigger()
  frame = await fg.wait_for_frame()
  cartesian_3d = frame.get_buffer(buffer_id.XYZ)
  fg.stop()

def worker():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    asyncio.ensure_future(bottom())
    asyncio.ensure_future(upper())
    loop.run_forever()
    
worker()

**Comment by dekhanra** [comment]
This is known issue with the Firmware version 0.16.23, where the camera has to be triggered two to four times to successfully acquire the frame. The State of the Port has to be changed from IDLE --> RUN --> IDLE before triggering the camera. The following code snippet gives anidea of working with swTrigger() via Python API.
```python
from ifm3dpy import FrameGrabber, buffer_id, O3R
import matplotlib.pyplot as plt

cam = O3R()

config = cam.to_json()

# Set the PORT state to IDLE before Triggering --> two state iterations required
for i in range(2):
    config['ports']['port2']['state'] = "RUN" #Expecting a head on Port 2
    cam.set(config)
    pcic_Port =  config['ports']['port2']['data']['pcicTCPPort']
    config['ports']['port2']['state'] = "IDLE" 
    cam.set(config)

fg = FrameGrabber(cam, pcic_Port)
fg.start([buffer_id.RADIAL_DISTANCE_IMAGE])

# Iterate the Software Trigger until the software trigger is accepted 
for _ in range(5):
    fg.sw_trigger()
    [ok, frame_port2] = fg.wait_for_frame().wait_for(200) # We estimate 100ms @ 10 Hz + some additional network delay/latency
    if ok:
        print('successfully the image is acquired')
        break
    else:
        print('Frame is not acquired in 200 milli Seconds. Triggering again')

rad_distance_image = frame_port2.get_buffer(buffer_id.RADIAL_DISTANCE_IMAGE)
plt.imshow(rad_distance_image)
plt.show()
```

This issue will be solved in the future releases.

If you have any further questions please let us know.

**Reply by Astronomia-maker** [reply]
Hi @dekhanra, I just want to ask you regarding software trigger mode bug. Is it already resolved in newest ifm3dpy==1.2.6 version?

**Reply by dekhanra** [reply]
Hi @Astronomia-maker,

Please find the information regarding the fixes and additions in our changelogs at https://api.ifm3d.com/html/content/ChangeLog.html

**Reply by desengph** [reply]
See information below

**Reply by BigBoot** [reply]
That change is not related to the sw trigger.
The sw trigger is a bug in the firmware of the device and can't be fixed by ifm3d.
As far as I know the bug is currently scheduled for the 1.1.x release of the firmware.

**Reply by desengph** [reply]
@BigBoot thank you for the clarification!

**Reply by Astronomia-maker** [reply]
Hi @BigBoot, @dekhanra, @desengph,

Thank you for your answers.

**Comment by Astronomia-maker** [comment]
I have just small question. i have two camera heads at port2 and port3. So my question regarding swtriggergroup value. both ports have different swtrigger value? How can i differentiate swtrigger group while data acquisition is on going. Because i am working mostly with 3D data and due multi camera interference, there are some noise points (flying pixels). So i want to avoid this noise using swtrigger. 

And at start of camera heads, would  it be in RUN mode or IDLE mode?

**Reply by dekhanra** [reply]
Thank you for your reply. 

Hey @Astronomia-maker, we have trouble in understanding your sentance structure. Could you please rephrase this _**How can i differentiate swtrigger group while data acquisition is on going**_

So, to mitigate the crosstalk between the cameras the difference between the channel values should be atleast or greater than 2. This has been already implemented in your JSON Schema.  

whereas the Software TriggerGroup is a parameter which allows you to trigger all the cameras in same group. In the config file you have mentioned the SWTriggerGroup for two camera heads on Port2 and Port3 are different which means you have to trigger both cameras independently to acquire the data.

And the **sw_trigger()** only triggers the camera if it is in IDLE State.

At the start, all the cameras will be in 'CONF' state as mentioned in Release Notes https://ifm3d.com/documentation/Firmware/ReleaseNotes/FW_0.16.x_Release_Notes.html

You can also reach us via mail robotics.support@ifm.com

**Reply by Astronomia-maker** [reply]
Sorry for sentence structure. Currently i have set both cameras at different channelvalue 2 and 4. And yes you are correct, i want to trigger both camera separately for 3d data acquisition. when first is acquiring data, second camera would be in IDLE state and vice versa.

**Comment by Astronomia-maker** [comment]
Hi, I tried followed python code and i am getting 3D data 1 in 5 execution. I tried to execute code and sometimes i have output only from 1 sensor or sometimes from another camera. Do you have any suggestion?

o3r = O3R()
config = o3r.to_json()

for i in range(2):
    config['ports']['port2']['state'] = "RUN" 
    config['ports']['port3']['state'] = "RUN" #Expecting a head on Port 2
    o3r.set(config)
    pcic_Port2 =  config['ports']['port2']['data']['pcicTCPPort']
    pcic_Port3 =  config['ports']['port3']['data']['pcicTCPPort']
    config['ports']['port2']['state'] = "IDLE" 
    config['ports']['port3']['state'] = "IDLE" 
    o3r.set(config)

fg_try = FrameGrabber(o3r, pcic_Port2)
fg_try.start([buffer_id.XYZ])

for _ in range(5):
    fg_try.sw_trigger()
    [ok, frame_port2] = fg_try.wait_for_frame().wait_for(200) # We estimate 100ms @ 10 Hz + some additional network delay/latency
    if ok:
        print('successfully the image is acquired')
        break
    else:
        print('Frame is not acquired in 200 milli Seconds. Triggering again')

async def bottom():
        # Initialize the objects
        #o3r = O3R('10.100.100.31')
        fg2 = FrameGrabber(o3r, pcic_Port2)        
        # Set schema and start Grabber
        fg2.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
        fg2.sw_trigger()
        # Get a frame
        frame = await fg2.wait_for_frame()
        # Read the distance image and display a pixel in the center
        cartesian_3d = frame.get_buffer(buffer_id.XYZ)
        fg2.stop()
        
async def upper():
        # Initialize the objects
        #o3r = O3R('10.100.100.31')
        fg3 = FrameGrabber(o3r, pcic_Port3)        
        # Set schema and start Grabber
        fg3.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
        fg3.sw_trigger()
        # Get a frame
        frame = await fg3.wait_for_frame()
        # Read the distance image and display a pixel in the center
        cartesian_3d = frame.get_buffer(buffer_id.XYZ)
        fg3.stop()

async def run_both_async_funcs():
    while True:
        # run async function 1 first
        await bottom()
        # wait for async function 1 to finish before running async function 2
        await upper()

def worker1():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    asyncio.ensure_future(run_both_async_funcs())
    loop.run_forever()

worker1()

**Reply by dekhanra** [reply]
can you please format your input as code.

**Reply by Astronomia-maker** [reply]
```
o3r = O3R()
config = o3r.to_json()

for i in range(2):
      config['ports']['port2']['state'] = "RUN"
      config['ports']['port3']['state'] = "RUN" #Expecting a head on Port 2
      o3r.set(config)
      pcic_Port2 = config['ports']['port2']['data']['pcicTCPPort']
      pcic_Port3 = config['ports']['port3']['data']['pcicTCPPort']
      config['ports']['port2']['state'] = "IDLE"
      config['ports']['port3']['state'] = "IDLE"
      o3r.set(config)

fg_try = FrameGrabber(o3r, pcic_Port2)
fg_try.start([buffer_id.XYZ])

for _ in range(5):
      fg_try.sw_trigger()
      [ok, frame_port2] = fg_try.wait_for_frame().wait_for(200) # We estimate 100ms @ 10 Hz + some additional network delay/latency
      if ok:
           print('successfully the image is acquired')
           break
      else:
          print('Frame is not acquired in 200 milli Seconds. Triggering again')

async def bottom():
        fg2 = FrameGrabber(o3r, pcic_Port2)
        fg2.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
        fg2.sw_trigger()
        frame = await fg2.wait_for_frame()
        cartesian_3d = frame.get_buffer(buffer_id.XYZ)
        fg2.stop()

async def upper():
        fg3 = FrameGrabber(o3r, pcic_Port3)
        fg3.start([buffer_id.NORM_AMPLITUDE_IMAGE,buffer_id.RADIAL_DISTANCE_IMAGE,buffer_id.XYZ])
        fg3.sw_trigger()
        frame = await fg3.wait_for_frame()
        cartesian_3d = frame.get_buffer(buffer_id.XYZ)
        fg3.stop()

async def run_both_async_funcs():
        while True:
                 await bottom()
                 await upper()

def worker1():
          loop = asyncio.new_event_loop()
          asyncio.set_event_loop(loop)
          asyncio.ensure_future(run_both_async_funcs())
          loop.run_forever()

worker1()
```

**Reply by desengph** [reply]
@Astronomia-maker please use simple Markdown syntax to add code: https://www.markdownguide.org/basic-syntax/#code-blocks-1

We tried to make sense of your code snippet here, but it seems you didn't include any of the suggestions as provided by @dekhanra.
Additionally, your software creates new framegrabber objects per loop and starts / stops them per loop iteration. This doesn't conform with the API design. 

Only on software trigger command, data retrieval commands, etc. should be looped on.

Unfortunately, we can not debug your code / application outside our APIs functionality. For support on how to use asyncio methods, please refer to other online resources. 

BR
Philipp

**Reply by Astronomia-maker** [reply]
Hi @desengph, Thank you for drawing my attention regarding my mistakes in code. I made changes accordingly and now it works. Thank you @dekhanra for providing me a sw_trigger() code snippet.

(Previously i was thinking i needed to start every time framegrabber and when finish the operation, stop the grabber. But i was wrong.)

**Reply by Astronomia-maker** [reply]
Hi, can you give more insight about swtriggergroup? For example, I have two camera heads, for 3d acquisition port 2 and port3 is equipped. Both cameras have different swtriggergroup value respectively 1 and 2. So as i understood, whenever sw_trigger() called for port2, it will acquire data only from port2, am i right? Or should i also specify swtriggergroup value?

**Reply by dekhanra** [reply]
Hi @Astronomia-maker,

Your Interpretation is correct on swTriggerGroup. So the description of this parameter says

 **All ports that have the same value as swTriggerGroup define a group. If any port of the group is triggered, then all ports in the group are triggered. An exception is swTriggerGroup==0 then no other ports are triggered [default setting]**

If the value is different(except 0) then you have to trigger both the ports separately.

It would be great to know your use case/application so that we can give suggestions on the usage of swTriggerGroup.

**Comment by Astronomia-maker** [comment]
Hi @dekhanra, My application is to capture same scene with two 3d camera  (both mounted in same direction but at different height) and merge both pointclouds. But if both camera run at same time then multi camera interference occurs and increase unwanted noise in pointclouds. So my objective is run both camera parallel with different channelvalue and using software trigger for acquiring pointclouds.

**Comment by dekhanra** [comment]
Hi @Astronomia-maker,

It sounds like you are trying to acquire the data at same time from the both the cameras. You can try the following things which may work for your application.

- Channel Values for Port2 and Port3 [Expected Heads connected to PORT2 and PORT3] are 10, 20 
  - This mitigates the Crosstalk between the cameras

- SoftwareTrigger Group value --> same for both the ports
  - When you trigger any of the camera the other camera will also be triggered
- Create two framegrabber instances for both the ports and trigger any one of the PORT and you can acquire data from the both framegrabbers.

Please try this and let me know if this works

BR
Rasheed

**Reply by Astronomia-maker** [reply]
Hi @dekhanra,

Thank you for your suggestion. But Does same trigger group create multi camert inference? Because in your suggestion channel values are 10 and 20 so there is crosstalk. But same trigger group will not create any unwanted noise??

**Comment by desengph** [comment]
**Crosstalk mitigation**

For multi camera setups there are two possible routes to mitigate multi camera interference:
1. Distinct channel values: This sets each camera to its own unique frequency so **NO** interference is possible - noise levels are **EXACTLY** the same for each channel value and channel difference > 2.
2. Changing time bases between cameras:
	1. in RUN state (continuous trigger internally): use the delay parameter to "offset" the respective cameras in time space: see link on [ifm3d.com](https://ifm3d.com/documentation/DeviceConfiguration/Parameters/AcquisitionSettings/triggering.html)

	2. in IDLE state (software trigger): trigger camera cameras consecutively with enough time "delay" to ensure no envelope signal overlap.

Based on the two (three) options above the simplest implementation is 1.
Choose option 2.1 if the cameras have to be synchronized - i.e. YOUR application.

For details on how to set each parameter see the documentation on [ifm3d.com](https://ifm3d.com/documentation/DeviceConfiguration/Parameters/AcquisitionSettings/triggering.html)

For details on the software implementation for option 1 / 2.1 see the suggestions given earlier (this QA thread).

I hope this clears things up. If you have additional questions about the principles discussed here, please implement the different options and benchmark them independently.

BR
Philipp
